### ISSUE TITLE
[Bug report] Undefined symbols for architecture x86_64 on Xcode build

### ISSUE THREAD

**Issue by douglasjunior** [realm/realm-js#3221]
## Goals

We are getting several errors `Undefined symbols for architecture x86_64` when try to build (play) the React Native app with Xcode for iOS.

Here I report the example of two apps, with two different versions of Realm and React Native, both with the same problem.

To solve, we need to:

```sh
cd ios
rm -rf Pods
pod install
```

This happens several times a day, we have not yet discovered the reason.

My guess is that this is related to branch exchange. However, when changing branches, neither React Native nor Realm versions are changed. (`ios/Pods` and `node_modules` are ignored on GIT)

### React Native 0.62.2 and RealmJS 6.0.2

![image](https://user-images.githubusercontent.com/1512341/93211454-a6af7500-f737-11ea-86ab-32864dba3af4.png)

### React Native 0.61.5 and RealmJS 3.6.5

![image](https://user-images.githubusercontent.com/1512341/93211512-b8911800-f737-11ea-8549-55ca15045842.png)

## Steps to Reproduce

~~Unfortunately, we still haven't found a standard behavior.~~

As @somebody32 says here https://github.com/realm/realm-js/issues/3221#issuecomment-723875769

> Looks like it is happening every time there is a change in Pods (ie adding a new or upgrading a version of any pod, not Realm). Realm is 10.0.1.
> 
> The temporal solution is to clean the build and do pod install again (no need to remove pods or node_modules folder), but I agree that it is pretty time consuming

## Code Sample

Both projects are created using `react-native init`, without any customization in the `metro`, `babel` or` cocoapods` scripts.

## Version of Realm and Tooling

- Realm JS SDK Version: 3.6.5 and 6.0.2
- React Native: RN 0.61.5 and 0.62.2 both with cocoapods and autolinking 
- Cocoapods: 1.9.3
- Xcode: 11.7 (11E801a)
- Node: 12.16.3 with nvm
- Client OS & Version: macOS Catalina 10.15.6 (19G2021)
- Which debugger for React Native: None

Maybe related to https://github.com/realm/realm-js/issues/2271 https://github.com/realm/realm-cocoa/issues/2393

**Comment by kraenhansen** [comment]
Are you comitting your `ios/Pods` directory into Git?
Does it help to clean out the build folder? (either by `rm -rf ios/build` or via the Xcode UI?

**Comment by douglasjunior** [comment]
Thanks for the quick reply!

The `ios/Pods` folder is ignored on GIT. 

We use the build only within Xcode, we never use it with "react-native run-ios", so there is no `ios/build` folder.

Before we found out that the problem was only in Pods, our solution was:

1. Xcode -> Product -> Clean
1. Remove Xcode `DerivedData`.
1. Remove `node_modules`
1. Remove `ios/Pods`

**Comment by kraenhansen** [comment]
So it works now?

**Comment by douglasjunior** [comment]
Yes, as I said, removing the Pods folder and installing again temporarily resolves the issue.

But, we still don't know the reason, the problem appears again, in different projects, different versions, different machines.

**Comment by whalemare** [comment]
I had this error after update from XCode 11.5 to XCode 12. Before this step, all works fine

**Comment by neopit** [comment]
I have the same error with realm@v10.0.0-beta.13.
Xcode version 11.7  and RN 0.63.2
Pretty annoying impossible to build/run the app

**Comment by hengkx** [comment]
I have the same error with realm@v6.12
Xcode version 11.7 and RN 0.63.2

**Comment by whalemare** [comment]
I wrote some hack for solve this issue for me for a while 
We need clear build folder and pods, as @douglasjunior says. So, we can do it in `postinstall` step

`bin/postInstall`
```js
#!/usr/bin/env node

const childProcess = require('child_process')
const os = require('os')
;[
  // Kill the metro bundler if it's running.
  { command: 'pkill -f "cli.js start" || set exit 0', onlyPlatforms: ['darwin', 'linux'] },

  // on iOS, make sure our native modules are installed
  { command: 'xcodebuild clean', cwd: 'ios', onlyPlatforms: ['darwin'] },
  {
    command: 'rm -rf ~/Library/Developer/Xcode/DerivedData/*',
    cwd: 'ios',
    onlyPlatforms: ['darwin'],
  },
  { command: 'pod install', cwd: 'ios', onlyPlatforms: ['darwin'] },
]
  .filter(({ onlyPlatforms }) => !onlyPlatforms || onlyPlatforms.includes(os.platform()))
  .forEach(commandAndOptions => {
    const { command, onlyPlatform: _, ...options } = commandAndOptions
    try {
      childProcess.execSync(command, {
        stdio: 'inherit',
        ...options,
      })
    } catch (error) {
      process.exit(error.status)
    }
  })
```

`package.json`
```jsonc
// other stuff
  "scripts": {
    "preinstall": "rm -rf node_modules && rm -rf ios/Pods",
    "postinstall": "node ./bin/postInstall"
   }
// other stuff
```

After this, I run project directly from XCode 12 and build success.

**Comment by somebody32** [comment]
Looks like it is happening every time there is a change in Pods (ie adding a new or upgrading a version of any pod, not Realm). 

The temporal solution is to clean the build and do pod install again (no need to remove pods or node_modules folder), but I agree that it is pretty time consuming

**Comment by awkpagong** [comment]
+1 same issue here every time my pods change... not very ideal

**Comment by douglasjunior** [comment]
Please guys, vote on the issue to increase its relevance (👍 ), so we can have attention from the Realm team. Adding random comments does not help the visibility of the issue.

**Comment by fronck** [comment]
We do have a fix in the pipeline ([PR-3221](https://github.com/realm/realm-js/pull/3455)), which should fix the issues when building for x86_64 on Xcode 12.
It's scheduled for the next Realm-JS release.
In the meantime, can you try building from the current repo, and let us know if it works for you?

**Comment by whalemare** [comment]
It's totally broke my development process when I need install new library.
Realm ftw is going on? Any comments here?

**Comment by kneth** [comment]
You can add something like 

```
post_install do |installer|
  installer.pods_project.build_configurations.each do |config|
    config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
  end
end
```

to your app's pod. See also https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios

**Comment by douglasjunior** [comment]
@kneth thanks for the tip, we will try asap.

However, our problem also happens in Xcode 11 and using a physical device, not just a simulator.

**Comment by sbrighiu** [comment]
Any progress? This is making my RN upgrade flow much more time consuming than it should be :(

**Comment by kneth** [comment]
@sbrighiu Did you try to modify your iOS app's Podfile as outlined [previously](https://github.com/realm/realm-js/issues/3221#issuecomment-755302234)?

**Comment by douglasjunior** [comment]
> You can add something like
> 
> ```
> post_install do |installer|
>   installer.pods_project.build_configurations.each do |config|
>     config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
>   end
> end
> ```
> 
> to your app's pod. See also https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios

Just bringing feedback, we added this snippet to our `Podfile`, but the problem keeps happening.

**Comment by sbrighiu** [comment]
I did actually, the result is the same. The only thing that works is cleaning the project before I do a pod install (just to be safe i do it before i run too :) )

**Comment by noldidrita** [comment]
I've also been encountering this same problem for months now. Been happening on Realm V3, V6 and now V10, so it's definitely not something caused by some recent change. My solution has been the same, clean build, reinstall pods, but it's more time-consuming than it needs to be, since it happens every time I need to Pod Install, even if Realm itself doesn't change.

**Comment by sbrighiu** [comment]
I dont have the time to check but does this happen when realm is not using cocoapods?

**Comment by sbrighiu** [comment]
I also posted here, maybe related ... https://github.com/realm/realm-cocoa/issues/6069#issuecomment-770220016

**Comment by sbrighiu** [comment]
I'm reposting here just to cover all topics, 

Guys i think i may of had 2 versions of cocoapods ... i'm not 100% but I cleaned everything up and now im sure i have just one. This is me explaining how I found it https://superuser.com/a/1627412/1276003.

TL;DR - If you ever ran `sudo gem install cocoapods -n /usr/local/bin` you may have 2 cocoapods installed. 
For me whenever i would call pod --version it would say 1.10.1, but today, from nowhere it said 1.8.4. Just run uninstall in both local and global (reference link above) and now I can run my `yarn` without cleaning 100 times.

I hope this helps someone. Please remember to only keep one cocoapods version. :)

**Comment by douglasjunior** [comment]
Here we have cocopoads installed with `homebrew` only, we never install with `gem install`.

**Comment by sbrighiu** [comment]
I just went in a day from 10 min debug/release build to 3 min release build and 2 min debug build by fixing this and removing Flipper from my project :) so I'm happy :) 

This fixes the issue I had with Xcode randomly showing 100 errors and stopping in the last 1%. I hope it helps someone :)

**Comment by sbrighiu** [comment]
Ok I had it happen again .. when that does I simply clean, delete derived data, run `yarn clean`, and then `yarn`
`
"clean": "rm -rf node_modules && rm -f package-lock.json && rm -f yarn.lock && rm -rf ios/Pods && rm -rf ios/Podfile.lock",
`

It is less frequent than before .. and not so much of an annoyance, but still not fixed...

**Comment by garmoshka-mo** [comment]
The most performant solution is to remove single Pod:
```
rm -rf Pods/Target\ Support\ Files/RealmJS
```

When you clean build, reinstall pods (pod deintegrate, rm -rf ~/Library/Caches/CocoaPods, rm -rf Pods) - then it takes much more time for new build.

**Comment by sbrighiu** [comment]
Yeah... I know, but when i do that i still need to do clean project (because sometimes something is just off, I can't figure out why the simple solution does not always work), yarn, pod install (just to have everything ok), jetify, etc...

Either way, my builds still take 10 minutes, so i got used to cleaning before i update the dependencies and if something happens, clean everything, delete derived, and since everything is cached locally in yarn/npm cache, it's just a difference of 1-2 minutes to get starting with release build.

Very... very... annoying :) I was thinking to maybe clean the RealmJS folder manually before I do the pod install, but i didn't have time to add that command back to my post-install hook.

I did this before, and when I saw i just need to clean before i do anything, I removed it :))

--------------------------------------
Stefan Brighiu
Senior Software Engineer @ Atta Systems
Phone: +40 744 581 921
Web: https://atta.systems

**Comment by douglasjunior** [comment]
Here we just need to:

```sh
cd ios/
rm -rf Pods/
pod install
```

No project clean, no cache clean, just reinstall pods, close and open Xcode, build.

**Comment by whalemare** [comment]
Why it is closed?
Reinstall pods take time that is unnecessary. Need to find issue that cause crash on build

**Comment by takameyer** [comment]
@whalemare We have updated our library to be built with a precompiled xcframework, which should prevent such crashes from occurring and also result in much faster builds.  Which version of Realm are you using?  Can you verify what steps you are taking to cause the crash?

**Comment by whalemare** [comment]
I am frozed my app with realm for now, just be wondering that issue closed without comments. 
I will try it in fresh version and write here if it's reproduced. 
Thank you
